gamemodes/terrortown/gamemode/server/sv_roleselection.lua

```diff
@@ -592,19 +607,25 @@ end
 -- @param table plys The players that should receive roles.
 -- @param number subrole The @{ROLE} index of the considered role.
 -- @param table selectableRoles The list of filtered selectable @{ROLE}s
+-- @param table selectedForcedRoles The List with a count of forced @{ROLE}s
 -- @realm server
 -- @internal
-local function UpgradeRoles(plys, subrole, selectableRoles)
+local function UpgradeRoles(plys, subrole, selectableRoles, selectedForcedRoles)
 	local availableRoles = {}
+	local roleAmount = 0
 
 	-- now upgrade this role if there are other subroles
 	for sub in pairs(selectableRoles) do
-		if roles.GetByIndex(sub).baserole == subrole then
+		roleAmount = selectableRoles[sub]
+		if selectedForcedRoles[sub] then
+			roleAmount = roleAmount - selectedForcedRoles[sub]
+		end
+		if roles.GetByIndex(sub).baserole == subrole and roleAmount > 0 then
 			availableRoles[#availableRoles + 1] = sub
 		end
 	end
 
-	SetSubRoles(plys, availableRoles, selectableRoles)
+	SetSubRoles(plys, availableRoles, selectableRoles, selectedForcedRoles)
```

Comment:
Then I guess you like this more general approach better. Subroles are now also being deducted by forced Roles.
We trade efficiency vs. reusability. The selectedForcedRoles can be an empty list and the selectableRoles are no longer being affected by calculations inside these functions.